// @polkadot/ui-settings/cjs/defaults/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UITHEMES = exports.UITHEME_DEFAULT = exports.UIMODES = exports.UIMODE_DEFAULT = exports.STORAGE_DEFAULT = exports.STORAGE = exports.PREFIXES = exports.PREFIX_DEFAULT = exports.NOTIFICATION_DEFAULT = exports.METADATA_UP_DEFAULT = exports.METADATA_UP = exports.LOCKING = exports.LOCKING_DEFAULT = exports.LEDGER_CONN = exports.LEDGER_CONN_DEFAULT = exports.LANGUAGE_DEFAULT = exports.ICONS = exports.ICON_DEFAULT_HOST = exports.ICON_DEFAULT = exports.ENDPOINTS = exports.ENDPOINT_DEFAULT = exports.CRYPTOS_LEDGER = exports.CRYPTOS_ETH = exports.CRYPTOS = exports.CAMERA = exports.CAMERA_DEFAULT = void 0;
const crypto_js_1 = require("./crypto.js");
Object.defineProperty(exports, "CRYPTOS", { enumerable: true, get: function () { return crypto_js_1.CRYPTOS; } });
Object.defineProperty(exports, "CRYPTOS_ETH", { enumerable: true, get: function () { return crypto_js_1.CRYPTOS_ETH; } });
Object.defineProperty(exports, "CRYPTOS_LEDGER", { enumerable: true, get: function () { return crypto_js_1.CRYPTOS_LEDGER; } });
const endpoints_js_1 = require("./endpoints.js");
Object.defineProperty(exports, "ENDPOINT_DEFAULT", { enumerable: true, get: function () { return endpoints_js_1.ENDPOINT_DEFAULT; } });
Object.defineProperty(exports, "ENDPOINTS", { enumerable: true, get: function () { return endpoints_js_1.ENDPOINTS; } });
const ledger_js_1 = require("./ledger.js");
Object.defineProperty(exports, "LEDGER_CONN", { enumerable: true, get: function () { return ledger_js_1.LEDGER_CONN; } });
Object.defineProperty(exports, "LEDGER_CONN_DEFAULT", { enumerable: true, get: function () { return ledger_js_1.LEDGER_CONN_DEFAULT; } });
const ss58_js_1 = require("./ss58.js");
Object.defineProperty(exports, "PREFIX_DEFAULT", { enumerable: true, get: function () { return ss58_js_1.PREFIX_DEFAULT; } });
Object.defineProperty(exports, "PREFIXES", { enumerable: true, get: function () { return ss58_js_1.PREFIXES; } });
const ui_js_1 = require("./ui.js");
Object.defineProperty(exports, "ICON_DEFAULT", { enumerable: true, get: function () { return ui_js_1.ICON_DEFAULT; } });
Object.defineProperty(exports, "ICON_DEFAULT_HOST", { enumerable: true, get: function () { return ui_js_1.ICON_DEFAULT_HOST; } });
Object.defineProperty(exports, "ICONS", { enumerable: true, get: function () { return ui_js_1.ICONS; } });
Object.defineProperty(exports, "NOTIFICATION_DEFAULT", { enumerable: true, get: function () { return ui_js_1.NOTIFICATION_DEFAULT; } });
Object.defineProperty(exports, "UIMODE_DEFAULT", { enumerable: true, get: function () { return ui_js_1.UIMODE_DEFAULT; } });
Object.defineProperty(exports, "UIMODES", { enumerable: true, get: function () { return ui_js_1.UIMODES; } });
Object.defineProperty(exports, "UITHEME_DEFAULT", { enumerable: true, get: function () { return ui_js_1.UITHEME_DEFAULT; } });
Object.defineProperty(exports, "UITHEMES", { enumerable: true, get: function () { return ui_js_1.UITHEMES; } });
const CAMERA_DEFAULT = 'off';
exports.CAMERA_DEFAULT = CAMERA_DEFAULT;
const CAMERA = [
    {
        info: 'on',
        text: 'Allow camera access',
        value: 'on'
    },
    {
        info: 'off',
        text: 'Do not allow camera access',
        value: 'off'
    }
];
exports.CAMERA = CAMERA;
const LANGUAGE_DEFAULT = 'default';
exports.LANGUAGE_DEFAULT = LANGUAGE_DEFAULT;
const LOCKING_DEFAULT = 'session';
exports.LOCKING_DEFAULT = LOCKING_DEFAULT;
const LOCKING = [
    {
        info: 'session',
        text: 'Once per session',
        value: 'session'
    },
    {
        info: 'tx',
        text: 'On each transaction',
        value: 'tx'
    }
];
exports.LOCKING = LOCKING;
const METADATA_UP_DEFAULT = 'off';
exports.METADATA_UP_DEFAULT = METADATA_UP_DEFAULT;
const METADATA_UP = [
    {
        info: 'off',
        text: 'Do not auto-update extension metadata',
        value: 'off'
    },
    {
        info: 'on',
        text: 'Auto-update extension metadata',
        value: 'on'
    }
];
exports.METADATA_UP = METADATA_UP;
const STORAGE_DEFAULT = 'off';
exports.STORAGE_DEFAULT = STORAGE_DEFAULT;
const STORAGE = [
    {
        info: 'on',
        text: 'Allow local in-browser account storage',
        value: 'on'
    },
    {
        info: 'off',
        text: 'Do not allow local in-browser account storage',
        value: 'off'
    }
];
exports.STORAGE = STORAGE;
