// @gear-js/api/errors/program.errors.js
class SubmitProgramError extends Error {
    name = 'SubmitProgramError';
    constructor(message) {
        super(message || 'Unable to submit the program. Params are invalid');
    }
}
class ProgramDoesNotExistError extends Error {
    name = 'ProgramDoesNotExist';
    constructor(id) {
        super(`Program with id ${id} does not exist`);
    }
}
class PausedProgramDoesNotExistError extends Error {
    name = 'PausedProgramDoesNotExistError';
    constructor(id) {
        super(`Program with id ${id} not found in paused program storage`);
    }
}
class ProgramTerminatedError extends Error {
    name = 'ProgramTerminated';
    constructor(id) {
        super(`Program ${id} terminated`);
    }
}
class ProgramExitedError extends Error {
    name = 'ProgramExited';
    constructor(id) {
        super(`Program ${id} exited`);
    }
}
class CodeDoesNotExistError extends Error {
    name = 'CodeDoesNotExist';
    constructor(id) {
        super(`Code with id ${id} not found in the storage`);
    }
}
class ProgramHasNoMetahash extends Error {
    name = 'ProgramHasNoMetahash';
    constructor(id) {
        super(`Program with id ${id} has not metahash function`);
    }
}

export { CodeDoesNotExistError, PausedProgramDoesNotExistError, ProgramDoesNotExistError, ProgramExitedError, ProgramHasNoMetahash, ProgramTerminatedError, SubmitProgramError };
