// @gear-js/vara-ui/dist/vara-ui.js
import * as De from "react";
import Ye, { forwardRef as Y, useId as Z, useState as fe, useEffect as qe, useCallback as Cr } from "react";
import { createPortal as kr } from "react-dom";
var ue = { exports: {} }, V = {};
/**
 * @license React
 * react-jsx-runtime.production.min.js
 *
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
var Ae;
function Sr() {
  if (Ae)
    return V;
  Ae = 1;
  var l = Ye, s = Symbol.for("react.element"), o = Symbol.for("react.fragment"), n = Object.prototype.hasOwnProperty, v = l.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED.ReactCurrentOwner, h = { key: !0, ref: !0, __self: !0, __source: !0 };
  function p(m, d, j) {
    var b, E = {}, R = null, O = null;
    j !== void 0 && (R = "" + j), d.key !== void 0 && (R = "" + d.key), d.ref !== void 0 && (O = d.ref);
    for (b in d)
      n.call(d, b) && !h.hasOwnProperty(b) && (E[b] = d[b]);
    if (m && m.defaultProps)
      for (b in d = m.defaultProps, d)
        E[b] === void 0 && (E[b] = d[b]);
    return { $$typeof: s, type: m, key: R, ref: O, props: E, _owner: v.current };
  }
  return V.Fragment = o, V.jsx = p, V.jsxs = p, V;
}
var U = {};
/**
 * @license React
 * react-jsx-runtime.development.js
 *
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
var We;
function Tr() {
  return We || (We = 1, process.env.NODE_ENV !== "production" && function() {
    var l = Ye, s = Symbol.for("react.element"), o = Symbol.for("react.portal"), n = Symbol.for("react.fragment"), v = Symbol.for("react.strict_mode"), h = Symbol.for("react.profiler"), p = Symbol.for("react.provider"), m = Symbol.for("react.context"), d = Symbol.for("react.forward_ref"), j = Symbol.for("react.suspense"), b = Symbol.for("react.suspense_list"), E = Symbol.for("react.memo"), R = Symbol.for("react.lazy"), O = Symbol.for("react.offscreen"), Q = Symbol.iterator, ee = "@@iterator";
    function Be(e) {
      if (e === null || typeof e != "object")
        return null;
      var r = Q && e[Q] || e[ee];
      return typeof r == "function" ? r : null;
    }
    var A = l.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED;
    function C(e) {
      {
        for (var r = arguments.length, t = new Array(r > 1 ? r - 1 : 0), a = 1; a < r; a++)
          t[a - 1] = arguments[a];
        ze("error", e, t);
      }
    }
    function ze(e, r, t) {
      {
        var a = A.ReactDebugCurrentFrame, f = a.getStackAddendum();
        f !== "" && (r += "%s", t = t.concat([f]));
        var _ = t.map(function(u) {
          return String(u);
        });
        _.unshift("Warning: " + r), Function.prototype.apply.call(console[e], console, _);
      }
    }
    var He = !1, Je = !1, Ge = !1, Ke = !1, Xe = !1, de;
    de = Symbol.for("react.module.reference");
    function Ze(e) {
      return !!(typeof e == "string" || typeof e == "function" || e === n || e === h || Xe || e === v || e === j || e === b || Ke || e === O || He || Je || Ge || typeof e == "object" && e !== null && (e.$$typeof === R || e.$$typeof === E || e.$$typeof === p || e.$$typeof === m || e.$$typeof === d || // This needs to include all possible module reference object
      // types supported by any Flight configuration anywhere since
      // we don't know which Flight build this will end up being used
      // with.
      e.$$typeof === de || e.getModuleId !== void 0));
    }
    function Qe(e, r, t) {
      var a = e.displayName;
      if (a)
        return a;
      var f = r.displayName || r.name || "";
      return f !== "" ? t + "(" + f + ")" : t;
    }
    function _e(e) {
      return e.displayName || "Context";
    }
    function P(e) {
      if (e == null)
        return null;
      if (typeof e.tag == "number" && C("Received an unexpected object in getComponentNameFromType(). This is likely a bug in React. Please file an issue."), typeof e == "function")
        return e.displayName || e.name || null;
      if (typeof e == "string")
        return e;
      switch (e) {
        case n:
          return "Fragment";
        case o:
          return "Portal";
        case h:
          return "Profiler";
        case v:
          return "StrictMode";
        case j:
          return "Suspense";
        case b:
          return "SuspenseList";
      }
      if (typeof e == "object")
        switch (e.$$typeof) {
          case m:
            var r = e;
            return _e(r) + ".Consumer";
          case p:
            var t = e;
            return _e(t._context) + ".Provider";
          case d:
            return Qe(e, e.render, "ForwardRef");
          case E:
            var a = e.displayName || null;
            return a !== null ? a : P(e.type) || "Memo";
          case R: {
            var f = e, _ = f._payload, u = f._init;
            try {
              return P(u(_));
            } catch {
              return null;
            }
          }
        }
      return null;
    }
    var L = Object.assign, q = 0, ve, be, he, pe, ge, me, ye;
    function xe() {
    }
    xe.__reactDisabledLog = !0;
    function er() {
      {
        if (q === 0) {
          ve = console.log, be = console.info, he = console.warn, pe = console.error, ge = console.group, me = console.groupCollapsed, ye = console.groupEnd;
          var e = {
            configurable: !0,
            enumerable: !0,
            value: xe,
            writable: !0
          };
          Object.defineProperties(console, {
            info: e,
            log: e,
            warn: e,
            error: e,
            group: e,
            groupCollapsed: e,
            groupEnd: e
          });
        }
        q++;
      }
    }
    function rr() {
      {
        if (q--, q === 0) {
          var e = {
            configurable: !0,
            enumerable: !0,
            writable: !0
          };
          Object.defineProperties(console, {
            log: L({}, e, {
              value: ve
            }),
            info: L({}, e, {
              value: be
            }),
            warn: L({}, e, {
              value: he
            }),
            error: L({}, e, {
              value: pe
            }),
            group: L({}, e, {
              value: ge
            }),
            groupCollapsed: L({}, e, {
              value: me
            }),
            groupEnd: L({}, e, {
              value: ye
            })
          });
        }
        q < 0 && C("disabledDepth fell below zero. This is a bug in React. Please file an issue.");
      }
    }
    var re = A.ReactCurrentDispatcher, te;
    function H(e, r, t) {
      {
        if (te === void 0)
          try {
            throw Error();
          } catch (f) {
            var a = f.stack.trim().match(/\n( *(at )?)/);
            te = a && a[1] || "";
          }
        return `
` + te + e;
      }
    }
    var ne = !1, J;
    {
      var tr = typeof WeakMap == "function" ? WeakMap : Map;
      J = new tr();
    }
    function Ee(e, r) {
      if (!e || ne)
        return "";
      {
        var t = J.get(e);
        if (t !== void 0)
          return t;
      }
      var a;
      ne = !0;
      var f = Error.prepareStackTrace;
      Error.prepareStackTrace = void 0;
      var _;
      _ = re.current, re.current = null, er();
      try {
        if (r) {
          var u = function() {
            throw Error();
          };
          if (Object.defineProperty(u.prototype, "props", {
            set: function() {
              throw Error();
            }
          }), typeof Reflect == "object" && Reflect.construct) {
            try {
              Reflect.construct(u, []);
            } catch (N) {
              a = N;
            }
            Reflect.construct(e, [], u);
          } else {
            try {
              u.call();
            } catch (N) {
              a = N;
            }
            e.call(u.prototype);
          }
        } else {
          try {
            throw Error();
          } catch (N) {
            a = N;
          }
          e();
        }
      } catch (N) {
        if (N && a && typeof N.stack == "string") {
          for (var c = N.stack.split(`
`), k = a.stack.split(`
`), g = c.length - 1, y = k.length - 1; g >= 1 && y >= 0 && c[g] !== k[y]; )
            y--;
          for (; g >= 1 && y >= 0; g--, y--)
            if (c[g] !== k[y]) {
              if (g !== 1 || y !== 1)
                do
                  if (g--, y--, y < 0 || c[g] !== k[y]) {
                    var T = `
` + c[g].replace(" at new ", " at ");
                    return e.displayName && T.includes("<anonymous>") && (T = T.replace("<anonymous>", e.displayName)), typeof e == "function" && J.set(e, T), T;
                  }
                while (g >= 1 && y >= 0);
              break;
            }
        }
      } finally {
        ne = !1, re.current = _, rr(), Error.prepareStackTrace = f;
      }
      var I = e ? e.displayName || e.name : "", Le = I ? H(I) : "";
      return typeof e == "function" && J.set(e, Le), Le;
    }
    function nr(e, r, t) {
      return Ee(e, !1);
    }
    function ar(e) {
      var r = e.prototype;
      return !!(r && r.isReactComponent);
    }
    function G(e, r, t) {
      if (e == null)
        return "";
      if (typeof e == "function")
        return Ee(e, ar(e));
      if (typeof e == "string")
        return H(e);
      switch (e) {
        case j:
          return H("Suspense");
        case b:
          return H("SuspenseList");
      }
      if (typeof e == "object")
        switch (e.$$typeof) {
          case d:
            return nr(e.render);
          case E:
            return G(e.type, r, t);
          case R: {
            var a = e, f = a._payload, _ = a._init;
            try {
              return G(_(f), r, t);
            } catch {
            }
          }
        }
      return "";
    }
    var K = Object.prototype.hasOwnProperty, Re = {}, je = A.ReactDebugCurrentFrame;
    function X(e) {
      if (e) {
        var r = e._owner, t = G(e.type, e._source, r ? r.type : null);
        je.setExtraStackFrame(t);
      } else
        je.setExtraStackFrame(null);
    }
    function sr(e, r, t, a, f) {
      {
        var _ = Function.call.bind(K);
        for (var u in e)
          if (_(e, u)) {
            var c = void 0;
            try {
              if (typeof e[u] != "function") {
                var k = Error((a || "React class") + ": " + t + " type `" + u + "` is invalid; it must be a function, usually from the `prop-types` package, but received `" + typeof e[u] + "`.This often happens because of typos such as `PropTypes.function` instead of `PropTypes.func`.");
                throw k.name = "Invariant Violation", k;
              }
              c = e[u](r, u, a, t, null, "SECRET_DO_NOT_PASS_THIS_OR_YOU_WILL_BE_FIRED");
            } catch (g) {
              c = g;
            }
            c && !(c instanceof Error) && (X(f), C("%s: type specification of %s `%s` is invalid; the type checker function must return `null` or an `Error` but returned a %s. You may have forgotten to pass an argument to the type checker creator (arrayOf, instanceOf, objectOf, oneOf, oneOfType, and shape all require an argument).", a || "React class", t, u, typeof c), X(null)), c instanceof Error && !(c.message in Re) && (Re[c.message] = !0, X(f), C("Failed %s type: %s", t, c.message), X(null));
          }
      }
    }
    var or = Array.isArray;
    function ae(e) {
      return or(e);
    }
    function lr(e) {
      {
        var r = typeof Symbol == "function" && Symbol.toStringTag, t = r && e[Symbol.toStringTag] || e.constructor.name || "Object";
        return t;
      }
    }
    function ir(e) {
      try {
        return Ce(e), !1;
      } catch {
        return !0;
      }
    }
    function Ce(e) {
      return "" + e;
    }
    function ke(e) {
      if (ir(e))
        return C("The provided key is an unsupported type %s. This value must be coerced to a string before before using it here.", lr(e)), Ce(e);
    }
    var M = A.ReactCurrentOwner, cr = {
      key: !0,
      ref: !0,
      __self: !0,
      __source: !0
    }, Se, Te, se;
    se = {};
    function ur(e) {
      if (K.call(e, "ref")) {
        var r = Object.getOwnPropertyDescriptor(e, "ref").get;
        if (r && r.isReactWarning)
          return !1;
      }
      return e.ref !== void 0;
    }
    function fr(e) {
      if (K.call(e, "key")) {
        var r = Object.getOwnPropertyDescriptor(e, "key").get;
        if (r && r.isReactWarning)
          return !1;
      }
      return e.key !== void 0;
    }
    function dr(e, r) {
      if (typeof e.ref == "string" && M.current && r && M.current.stateNode !== r) {
        var t = P(M.current.type);
        se[t] || (C('Component "%s" contains the string ref "%s". Support for string refs will be removed in a future major release. This case cannot be automatically converted to an arrow function. We ask you to manually fix this case by using useRef() or createRef() instead. Learn more about using refs safely here: https://reactjs.org/link/strict-mode-string-ref', P(M.current.type), e.ref), se[t] = !0);
      }
    }
    function _r(e, r) {
      {
        var t = function() {
          Se || (Se = !0, C("%s: `key` is not a prop. Trying to access it will result in `undefined` being returned. If you need to access the same value within the child component, you should pass it as a different prop. (https://reactjs.org/link/special-props)", r));
        };
        t.isReactWarning = !0, Object.defineProperty(e, "key", {
          get: t,
          configurable: !0
        });
      }
    }
    function vr(e, r) {
      {
        var t = function() {
          Te || (Te = !0, C("%s: `ref` is not a prop. Trying to access it will result in `undefined` being returned. If you need to access the same value within the child component, you should pass it as a different prop. (https://reactjs.org/link/special-props)", r));
        };
        t.isReactWarning = !0, Object.defineProperty(e, "ref", {
          get: t,
          configurable: !0
        });
      }
    }
    var br = function(e, r, t, a, f, _, u) {
      var c = {
        // This tag allows us to uniquely identify this as a React Element
        $$typeof: s,
        // Built-in properties that belong on the element
        type: e,
        key: r,
        ref: t,
        props: u,
        // Record the component responsible for creating this element.
        _owner: _
      };
      return c._store = {}, Object.defineProperty(c._store, "validated", {
        configurable: !1,
        enumerable: !1,
        writable: !0,
        value: !1
      }), Object.defineProperty(c, "_self", {
        configurable: !1,
        enumerable: !1,
        writable: !1,
        value: a
      }), Object.defineProperty(c, "_source", {
        configurable: !1,
        enumerable: !1,
        writable: !1,
        value: f
      }), Object.freeze && (Object.freeze(c.props), Object.freeze(c)), c;
    };
    function hr(e, r, t, a, f) {
      {
        var _, u = {}, c = null, k = null;
        t !== void 0 && (ke(t), c = "" + t), fr(r) && (ke(r.key), c = "" + r.key), ur(r) && (k = r.ref, dr(r, f));
        for (_ in r)
          K.call(r, _) && !cr.hasOwnProperty(_) && (u[_] = r[_]);
        if (e && e.defaultProps) {
          var g = e.defaultProps;
          for (_ in g)
            u[_] === void 0 && (u[_] = g[_]);
        }
        if (c || k) {
          var y = typeof e == "function" ? e.displayName || e.name || "Unknown" : e;
          c && _r(u, y), k && vr(u, y);
        }
        return br(e, c, k, f, a, M.current, u);
      }
    }
    var oe = A.ReactCurrentOwner, we = A.ReactDebugCurrentFrame;
    function W(e) {
      if (e) {
        var r = e._owner, t = G(e.type, e._source, r ? r.type : null);
        we.setExtraStackFrame(t);
      } else
        we.setExtraStackFrame(null);
    }
    var le;
    le = !1;
    function ie(e) {
      return typeof e == "object" && e !== null && e.$$typeof === s;
    }
    function $e() {
      {
        if (oe.current) {
          var e = P(oe.current.type);
          if (e)
            return `

Check the render method of \`` + e + "`.";
        }
        return "";
      }
    }
    function pr(e) {
      {
        if (e !== void 0) {
          var r = e.fileName.replace(/^.*[\\\/]/, ""), t = e.lineNumber;
          return `

Check your code at ` + r + ":" + t + ".";
        }
        return "";
      }
    }
    var Oe = {};
    function gr(e) {
      {
        var r = $e();
        if (!r) {
          var t = typeof e == "string" ? e : e.displayName || e.name;
          t && (r = `

Check the top-level render call using <` + t + ">.");
        }
        return r;
      }
    }
    function Pe(e, r) {
      {
        if (!e._store || e._store.validated || e.key != null)
          return;
        e._store.validated = !0;
        var t = gr(r);
        if (Oe[t])
          return;
        Oe[t] = !0;
        var a = "";
        e && e._owner && e._owner !== oe.current && (a = " It was passed a child from " + P(e._owner.type) + "."), W(e), C('Each child in a list should have a unique "key" prop.%s%s See https://reactjs.org/link/warning-keys for more information.', t, a), W(null);
      }
    }
    function Ne(e, r) {
      {
        if (typeof e != "object")
          return;
        if (ae(e))
          for (var t = 0; t < e.length; t++) {
            var a = e[t];
            ie(a) && Pe(a, r);
          }
        else if (ie(e))
          e._store && (e._store.validated = !0);
        else if (e) {
          var f = Be(e);
          if (typeof f == "function" && f !== e.entries)
            for (var _ = f.call(e), u; !(u = _.next()).done; )
              ie(u.value) && Pe(u.value, r);
        }
      }
    }
    function mr(e) {
      {
        var r = e.type;
        if (r == null || typeof r == "string")
          return;
        var t;
        if (typeof r == "function")
          t = r.propTypes;
        else if (typeof r == "object" && (r.$$typeof === d || // Note: Memo only checks outer props here.
        // Inner props are checked in the reconciler.
        r.$$typeof === E))
          t = r.propTypes;
        else
          return;
        if (t) {
          var a = P(r);
          sr(t, e.props, "prop", a, e);
        } else if (r.PropTypes !== void 0 && !le) {
          le = !0;
          var f = P(r);
          C("Component %s declared `PropTypes` instead of `propTypes`. Did you misspell the property assignment?", f || "Unknown");
        }
        typeof r.getDefaultProps == "function" && !r.getDefaultProps.isReactClassApproved && C("getDefaultProps is only used on classic React.createClass definitions. Use a static property named `defaultProps` instead.");
      }
    }
    function yr(e) {
      {
        for (var r = Object.keys(e.props), t = 0; t < r.length; t++) {
          var a = r[t];
          if (a !== "children" && a !== "key") {
            W(e), C("Invalid prop `%s` supplied to `React.Fragment`. React.Fragment can only have `key` and `children` props.", a), W(null);
            break;
          }
        }
        e.ref !== null && (W(e), C("Invalid attribute `ref` supplied to `React.Fragment`."), W(null));
      }
    }
    function Fe(e, r, t, a, f, _) {
      {
        var u = Ze(e);
        if (!u) {
          var c = "";
          (e === void 0 || typeof e == "object" && e !== null && Object.keys(e).length === 0) && (c += " You likely forgot to export your component from the file it's defined in, or you might have mixed up default and named imports.");
          var k = pr(f);
          k ? c += k : c += $e();
          var g;
          e === null ? g = "null" : ae(e) ? g = "array" : e !== void 0 && e.$$typeof === s ? (g = "<" + (P(e.type) || "Unknown") + " />", c = " Did you accidentally export a JSX literal instead of a component?") : g = typeof e, C("React.jsx: type is invalid -- expected a string (for built-in components) or a class/function (for composite components) but got: %s.%s", g, c);
        }
        var y = hr(e, r, t, f, _);
        if (y == null)
          return y;
        if (u) {
          var T = r.children;
          if (T !== void 0)
            if (a)
              if (ae(T)) {
                for (var I = 0; I < T.length; I++)
                  Ne(T[I], e);
                Object.freeze && Object.freeze(T);
              } else
                C("React.jsx: Static children should always be an array. You are likely explicitly calling React.jsxs or React.jsxDEV. Use the Babel transform instead.");
            else
              Ne(T, e);
        }
        return e === n ? yr(y) : mr(y), y;
      }
    }
    function xr(e, r, t) {
      return Fe(e, r, t, !0);
    }
    function Er(e, r, t) {
      return Fe(e, r, t, !1);
    }
    var Rr = Er, jr = xr;
    U.Fragment = n, U.jsx = Rr, U.jsxs = jr;
  }()), U;
}
process.env.NODE_ENV === "production" ? ue.exports = Sr() : ue.exports = Tr();
var i = ue.exports;
function Me(l) {
  var s, o, n = "";
  if (typeof l == "string" || typeof l == "number")
    n += l;
  else if (typeof l == "object")
    if (Array.isArray(l))
      for (s = 0; s < l.length; s++)
        l[s] && (o = Me(l[s])) && (n && (n += " "), n += o);
    else
      for (s in l)
        l[s] && (n && (n += " "), n += s);
  return n;
}
function x() {
  for (var l, s, o = 0, n = ""; o < arguments.length; )
    (l = arguments[o++]) && (s = Me(l)) && (n && (n += " "), n += s);
  return n;
}
const wr = "_button_tg87l_9", $r = "_disabled_tg87l_28", Or = "_loading_tg87l_31", Pr = "_blink_tg87l_1", Nr = "_block_tg87l_36", Fr = "_noWrap_tg87l_40", Lr = "_primary_tg87l_44", Dr = "_dark_tg87l_56", Ar = "_light_tg87l_69", Wr = "_grey_tg87l_80", Ir = "_border_tg87l_91", Yr = "_transparent_tg87l_104", qr = "_noText_tg87l_115", Mr = "_small_tg87l_120", F = {
  button: wr,
  disabled: $r,
  loading: Or,
  blink: Pr,
  block: Nr,
  noWrap: Fr,
  primary: Lr,
  dark: Dr,
  light: Ar,
  grey: Wr,
  border: Ir,
  transparent: Yr,
  default: "_default_tg87l_111",
  noText: qr,
  small: Mr
}, Ve = Y((l, s) => {
  const {
    className: o,
    text: n,
    icon: v,
    disabled: h,
    isLoading: p,
    type: m = "button",
    color: d = "primary",
    size: j = "default",
    children: b,
    block: E,
    noWrap: R,
    ...O
  } = l;
  return /* @__PURE__ */ i.jsxs(
    "button",
    {
      type: m,
      className: x(
        F.button,
        F[d],
        d !== "transparent" && F[j],
        h && F.disabled,
        p && F.loading,
        !n && !b && F.noText,
        E && F.block,
        R && F.noWrap,
        o
      ),
      disabled: h || p,
      ref: s,
      ...O,
      children: [
        v && /* @__PURE__ */ i.jsx(v, {}),
        n && /* @__PURE__ */ i.jsx("span", { children: n }),
        b
      ]
    }
  );
}), Vr = "_root_cpr6h_1", Ur = "_disabled_cpr6h_1", Br = "_base_cpr6h_6", zr = "_small_cpr6h_15", Hr = "_input_cpr6h_19", Jr = "_label_cpr6h_20", Gr = "_fieldset_cpr6h_57", Kr = "_legendLabel_cpr6h_57", Xr = "_error_cpr6h_68", Zr = "_legend_cpr6h_57", Qr = "_message_cpr6h_101", S = {
  root: Vr,
  disabled: Ur,
  base: Br,
  default: "_default_cpr6h_12",
  small: zr,
  input: Hr,
  label: Jr,
  fieldset: Gr,
  legendLabel: Kr,
  error: Xr,
  legend: Zr,
  message: Qr
};
function et() {
  const [l, s] = fe(0), o = Z();
  return qe(() => {
    const n = document.getElementById(o);
    n && s(n.getBoundingClientRect().width);
  }, [o]), [l, o];
}
const Jt = Y(
  ({ icon: l, className: s, label: o, error: n, type: v = "text", placeholder: h = " ", size: p = "default", ...m }, d) => {
    const { disabled: j } = m, b = Z(), [E, R] = et(), O = 14, ee = { left: `${E ? O + 8 + E : O}px` };
    return /* @__PURE__ */ i.jsxs("div", { className: x(S.root, s, j && S.disabled), children: [
      /* @__PURE__ */ i.jsxs("div", { className: x(S.base, S[p]), children: [
        l && /* @__PURE__ */ i.jsx(l, { id: R }),
        /* @__PURE__ */ i.jsx(
          "input",
          {
            type: v,
            id: b,
            className: x(S.input, S[p], n && S.error),
            placeholder: h,
            ref: d,
            ...m
          }
        ),
        o && /* @__PURE__ */ i.jsx("label", { htmlFor: b, className: x(S.label, S[p]), style: ee, children: o }),
        /* @__PURE__ */ i.jsx("fieldset", { className: S.fieldset, children: /* @__PURE__ */ i.jsxs("legend", { className: x(S.legend, o && S.legendLabel), children: [
          o,
          "​"
        ] }) })
      ] }),
      n && /* @__PURE__ */ i.jsx("p", { className: S.message, children: n })
    ] });
  }
), rt = "_root_1f7h4_1", tt = "_disabled_1f7h4_1", nt = "_base_1f7h4_6", at = "_textarea_1f7h4_10", st = "_label_1f7h4_11", ot = "_small_1f7h4_21", lt = "_error_1f7h4_38", it = "_fieldset_1f7h4_41", ct = "_legendLabel_1f7h4_41", ut = "_legend_1f7h4_41", ft = "_message_1f7h4_95", w = {
  root: rt,
  disabled: tt,
  base: nt,
  textarea: at,
  label: st,
  default: "_default_1f7h4_17",
  small: ot,
  error: lt,
  fieldset: it,
  legendLabel: ct,
  legend: ut,
  message: ft
}, Gt = Y(
  ({ className: l, label: s, error: o, size: n = "default", rows: v = 5, placeholder: h = " ", ...p }, m) => {
    const { disabled: d } = p, j = Z();
    return /* @__PURE__ */ i.jsxs("div", { className: x(w.root, l, d && w.disabled), children: [
      /* @__PURE__ */ i.jsxs("div", { className: w.base, children: [
        /* @__PURE__ */ i.jsx(
          "textarea",
          {
            rows: v,
            id: j,
            className: x(w.textarea, w[n], o && w.error),
            placeholder: h,
            ref: m,
            ...p
          }
        ),
        s && /* @__PURE__ */ i.jsx("label", { htmlFor: j, className: x(w.label, w[n]), children: s }),
        /* @__PURE__ */ i.jsx("fieldset", { className: w.fieldset, children: /* @__PURE__ */ i.jsxs("legend", { className: x(w.legend, s && w.legendLabel), children: [
          s,
          "​"
        ] }) })
      ] }),
      o && /* @__PURE__ */ i.jsx("p", { className: w.message, children: o })
    ] });
  }
), dt = "_root_kvfe0_1", _t = "_disabled_kvfe0_1", vt = "_base_kvfe0_6", bt = "_select_kvfe0_10", ht = "_label_kvfe0_11", pt = "_small_kvfe0_22", gt = "_error_kvfe0_39", mt = "_fieldset_kvfe0_42", yt = "_legendLabel_kvfe0_42", xt = "_legend_kvfe0_42", Et = "_message_kvfe0_91", $ = {
  root: dt,
  disabled: _t,
  base: vt,
  select: bt,
  label: ht,
  default: "_default_kvfe0_18",
  small: pt,
  error: gt,
  fieldset: mt,
  legendLabel: yt,
  legend: xt,
  message: Et
}, Kt = Y(
  ({ options: l, className: s, label: o, error: n, size: v = "default", ...h }, p) => {
    const { disabled: m } = h, d = Z(), j = () => l.map((b, E) => /* @__PURE__ */ i.jsx("option", { ...b }, E));
    return /* @__PURE__ */ i.jsxs("div", { className: x($.root, s, m && $.disabled), children: [
      /* @__PURE__ */ i.jsxs("div", { className: $.base, children: [
        /* @__PURE__ */ i.jsx("select", { id: d, className: x($.select, $[v], n && $.error), ref: p, ...h, children: j() }),
        o && /* @__PURE__ */ i.jsx("label", { htmlFor: d, className: x($.label, $[v]), children: o }),
        /* @__PURE__ */ i.jsx("fieldset", { className: $.fieldset, children: /* @__PURE__ */ i.jsxs("legend", { className: x($.legend, o && $.legendLabel), children: [
          o,
          "​"
        ] }) })
      ] }),
      n && /* @__PURE__ */ i.jsx("p", { className: $.message, children: n })
    ] });
  }
), Rt = "_label_1qqar_1", jt = "_disabled_1qqar_6", Ct = "_input_1qqar_11", kt = "_checkbox_1qqar_17", B = {
  label: Rt,
  disabled: jt,
  input: Ct,
  checkbox: kt,
  switch: "_switch_1qqar_26"
}, Xt = Y(({ label: l, className: s, type: o, ...n }, v) => {
  const { disabled: h } = n;
  return /* @__PURE__ */ i.jsxs("label", { className: x(B.label, s, h && B.disabled), children: [
    /* @__PURE__ */ i.jsx(
      "input",
      {
        type: "checkbox",
        className: x(B.input, o === "switch" ? B.switch : B.checkbox),
        ref: v,
        ...n
      }
    ),
    l
  ] });
}), St = "_label_e8vr5_1", Tt = "_disabled_e8vr5_6", wt = "_input_e8vr5_11", ce = {
  label: St,
  disabled: Tt,
  input: wt
}, Zt = Y(({ label: l, className: s, ...o }, n) => {
  const { disabled: v } = o;
  return /* @__PURE__ */ i.jsxs("label", { className: x(ce.label, s, v && ce.disabled), children: [
    /* @__PURE__ */ i.jsx("input", { type: "radio", className: ce.input, ref: n, ...o }),
    l
  ] });
}), Ue = (l) => /* @__PURE__ */ De.createElement("svg", { width: 16, height: 16, viewBox: "0 0 16 16", fill: "none", xmlns: "http://www.w3.org/2000/svg", ...l }, /* @__PURE__ */ De.createElement("path", { d: "M1.33454 0.54405C1.16856 0.544071 1.00638 0.593719 0.868832 0.686604C0.731282 0.779488 0.624641 0.911362 0.562608 1.06531C0.500576 1.21925 0.485983 1.38827 0.520706 1.55057C0.555429 1.71287 0.637881 1.86103 0.757471 1.97612L6.82396 8.04259L0.757471 14.1092C0.6774 14.1856 0.613391 14.2773 0.569192 14.3788C0.524994 14.4803 0.501494 14.5896 0.500069 14.7003C0.498644 14.811 0.519322 14.9209 0.560893 15.0235C0.602464 15.1261 0.664092 15.2195 0.742169 15.298C0.820245 15.3765 0.9132 15.4386 1.01559 15.4807C1.11798 15.5229 1.22775 15.5442 1.33847 15.5434C1.44919 15.5426 1.55864 15.5197 1.6604 15.476C1.76216 15.4324 1.85419 15.3689 1.93111 15.2892L8.00247 9.22113L14.0706 15.2892C14.148 15.3669 14.24 15.4286 14.3413 15.4707C14.4426 15.5128 14.5512 15.5345 14.6609 15.5346C14.7706 15.5347 14.8793 15.5132 14.9806 15.4712C15.082 15.4293 15.1741 15.3678 15.2517 15.2902C15.3292 15.2126 15.3908 15.1206 15.4327 15.0192C15.4746 14.9178 15.4962 14.8091 15.4961 14.6994C15.496 14.5897 15.4743 14.4812 15.4322 14.3799C15.3901 14.2786 15.3284 14.1866 15.2507 14.1092L9.18099 8.04259L15.2507 1.97612C15.3724 1.85924 15.4557 1.70819 15.4896 1.54289C15.5235 1.37759 15.5063 1.20584 15.4403 1.05052C15.3744 0.895208 15.2628 0.763621 15.1204 0.673164C14.978 0.582707 14.8114 0.537686 14.6428 0.54405C14.4266 0.550493 14.2213 0.64089 14.0706 0.796039L8.00247 6.86414L1.93111 0.796039C1.85355 0.716311 1.76079 0.652891 1.65833 0.609615C1.55586 0.566339 1.44576 0.544057 1.33454 0.54405Z", fill: "black" })), $t = "_root_1vzej_1", Ot = "_alert_1vzej_9", Pt = "_header_1vzej_16", Nt = "_success_1vzej_35", Ft = "_error_1vzej_42", Lt = "_info_1vzej_49", Dt = "_loading_1vzej_56", At = "_rotating_1vzej_1", Wt = "_button_1vzej_64", It = "_body_1vzej_71", z = {
  root: $t,
  alert: Ot,
  header: Pt,
  success: Nt,
  error: Ft,
  info: Lt,
  loading: Dt,
  rotating: At,
  button: Wt,
  body: It
};
function Qt({ alert: l, close: s }) {
  const { content: o, options: n } = l, { type: v, title: h, style: p, isClosed: m } = n;
  return /* @__PURE__ */ i.jsxs("div", { className: z.alert, style: p, children: [
    /* @__PURE__ */ i.jsxs("header", { className: x(z.header, z[v]), children: [
      h || v,
      m && /* @__PURE__ */ i.jsx(Ve, { icon: Ue, color: "transparent", className: z.button, onClick: s })
    ] }),
    /* @__PURE__ */ i.jsx("div", { className: z.body, children: o })
  ] });
}
const Yt = "_customScroll_ymry0_1", qt = "_overlay_ymry0_29", Mt = "_modal_ymry0_44", Vt = "_header_ymry0_60", Ut = "_heading_ymry0_65", D = {
  customScroll: Yt,
  overlay: qt,
  modal: Mt,
  header: Vt,
  heading: Ut
};
function Ie() {
  const [l, s] = fe(0), o = Cr((n) => {
    n && s(n.getBoundingClientRect().height);
  }, []);
  return [l, o];
}
function Bt() {
  const [l, s] = Ie(), [o, n] = Ie(), v = 32;
  return { bodyStyle: { maxHeight: `calc(100vh - ${l - o + 2 * v}px)` }, modalRef: s, bodyRef: n };
}
const en = ({ heading: l, close: s, children: o, className: n, footer: v }) => {
  const [h, p] = fe(), { bodyStyle: m, modalRef: d, bodyRef: j } = Bt(), b = ({ target: R, currentTarget: O }) => {
    R === O && s();
  };
  qe(() => {
    const R = document.createElement("div");
    return R.id = "modal-root", document.body.appendChild(R), p(R), () => {
      document.body.removeChild(R);
    };
  }, []);
  const E = /* @__PURE__ */ i.jsx("div", { className: D.overlay, onClick: b, children: /* @__PURE__ */ i.jsxs("div", { className: D.modal, ref: d, children: [
    /* @__PURE__ */ i.jsxs("header", { className: D.header, children: [
      /* @__PURE__ */ i.jsx("h3", { className: D.heading, children: l }),
      /* @__PURE__ */ i.jsx(Ve, { icon: Ue, color: "transparent", onClick: s, className: D.button })
    ] }),
    o && /* @__PURE__ */ i.jsx("div", { className: x(D.customScroll, n), style: m, ref: j, children: o }),
    v && /* @__PURE__ */ i.jsx("footer", { className: D.footer, children: v })
  ] }) });
  return h ? kr(E, h) : null;
};
export {
  Qt as Alert,
  Ve as Button,
  Xt as Checkbox,
  Jt as Input,
  en as Modal,
  Zt as Radio,
  Kt as Select,
  Gt as Textarea,
  z as alertStyles,
  F as buttonStyles,
  B as checkboxStyles,
  S as inputStyles,
  D as modalStyles,
  ce as radioStyles,
  $ as selectStyles,
  w as textareaStyles
};
